### PR TITLE
Added new tests for simultaneous use with mutable dispatch

### PR DESCRIPTION
**Pull request by shajder**
-cross queue simultaneous use
-in-order queue with simultaneous use

According to issue description #1481

### CODE REVIEW

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp [hunk 1]
```diff
@@ -22,29 +22,29 @@
...
         simultaneous_use_requested = simultaneous_request;
         if (simultaneous_request) buffer_size_multiplier = 2;
     }
 
-    //--------------------------------------------------------------------------
     cl_int SetUpKernel() override
     {
-        cl_int error = BasicMutableCommandBufferTest::SetUpKernel();
+        cl_int error = BasicCommandBufferTest::SetUpKernel();
         test_error(error, "BasicMutableCommandBufferTest::SetUpKernel failed");
```
Comment (EwanC) [inline test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp:48]:
```suggestion
        test_error(error, "BasicCommandBufferTest::SetUpKernel failed");
```

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp [hunk 2]
```diff
@@ -72,10 +72,9 @@ struct OutOfOrderTest : public BasicMutableCommandBufferTest
         return CL_SUCCESS;
     }
 
-    //--------------------------------------------------------------------------
     cl_int SetUpKernelArgs() override
     {
-        cl_int error = BasicMutableCommandBufferTest::SetUpKernelArgs();
+        cl_int error = BasicCommandBufferTest::SetUpKernelArgs();
         test_error(error,
                    "BasicMutableCommandBufferTest::SetUpKernelArgs failed");
```
Comment (EwanC) [inline test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp:79]:
```suggestion
                   "BasicCommandBufferTest::SetUpKernelArgs failed");
```

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp [hunk 3]
```diff
@@ -92,33 +91,39 @@ struct OutOfOrderTest : public BasicMutableCommandBufferTest
...
+                context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE,
+                &error);
+            test_error(error, "Unable to create command queue to test with");
+
+            cl_command_buffer_properties_khr properties[3] = {
+                CL_COMMAND_BUFFER_FLAGS_KHR, CL_COMMAND_BUFFER_MUTABLE_KHR, 0
+            };
+
+            work_command_buffer =
+                clCreateCommandBufferKHR(1, &work_queue, properties, &error);
```
Comment (EwanC) [inline test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp:111]: I think we need to set `CL_COMMAND_BUFFER_SIMULTANEOUS_USE_KHR` as well for `simultaneous_use_requested`, here and in the in-order queue case

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp [hunk 4]
```diff
@@ -429,26 +423,211 @@ struct OutOfOrderTest : public BasicMutableCommandBufferTest
...
+
+        error = clSetKernelArg(kernel, 1, sizeof(out_mem), &out_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int SetUp(int elements) override
+    {
+        cl_int error = BasicMutableCommandBufferTest::SetUp(elements);
```
Comment (EwanC) [inline test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp:485]: command-buffer created here might not have `CL_COMMAND_BUFFER_SIMULTANEOUS_USE_KHR` set on creation.

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp (hunk 1 above)
Comment (shajder) [reply test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp]: Done

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp (hunk 2 above)
Comment (shajder) [reply test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp]: Done

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp (hunk 4 above)
Comment (shajder) [reply test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_simultaneous.cpp:493]: Turns out base class for mutable dispatch subset of tests ignores `simultaneous_use_requested` flag - corrected as well

**Review by EwanC — APPROVED** [review]
LGTM - works for me with oneAPI Construction Kit, but did bump into an issue using the emulation layer that i've commented on.

File: test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_basic.h [hunk 5]
```diff
@@ -26,32 +26,50 @@ struct BasicMutableCommandBufferTest : BasicCommandBufferTest
...
+        cl_command_buffer_properties_khr prop = CL_COMMAND_BUFFER_MUTABLE_KHR;
+        if (simultaneous_use_support)
+        {
+            prop |= CL_COMMAND_BUFFER_SIMULTANEOUS_USE_KHR;
+        }
+
         const cl_command_buffer_properties_khr props[] = {
             CL_COMMAND_BUFFER_FLAGS_KHR,
-            CL_COMMAND_BUFFER_MUTABLE_KHR,
+            prop,
```
[the remark this answers is not in the document]
Comment (bashbaug) [reply test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_basic.h:66]: That's embarassing... I should get in the habit of running all the CTS tests with each commit, I suppose.

In any case, this is fixed now, and these tests are all passing with the layer also.

**Review by EwanC — APPROVED** [review]

**Review by bashbaug — APPROVED** [review]

**Pull request merged**